dist/tools/compile_commands/compile_commands.py

```diff
@@ -0,0 +1,260 @@
+#!/usr/bin/python3
+"""
+Command line utility to generate compile_commands.json for RIOT applications
+"""
+import argparse
+import os
+import shlex
+import subprocess
+import sys
+import re
+
+TEMPLATE_COMMAND = """  {{
+    "arguments": [
+      "{compiler}",
+      "-DRIOT_FILE_RELATIVE=\\\"{directory}/{source}\\\"",
+      "-DRIOT_FILE_NOPATH=\\\"{source}\\\"",
+{flags}      "-MQ",
+      "{output}",
+      "-MD",
+      "-MP",
+      "-c",
+      "-o",
+      "{output}",
+      "{source}"
+    ],
+    "directory": "{directory}",
+    "file": "{directory}/{source}",
+    "output": "{bin_directory}/{output}"
+  }}"""
+
+TEMPLATE_FLAG = "      \"{flag}\",\n"
+
+REGEX_VERSION = re.compile(r"\ngcc version ([^ ]+)")
+REGEX_INCLUDES = r"^#include <\.\.\.> search starts here:$((?:\n|\r|.)*?)^End of search list\.$"
+REGEX_INCLUDES = re.compile(REGEX_INCLUDES, re.MULTILINE)
+
+
+def detect_includes_and_version_gcc(compiler):
+    """
+    Runs the given compiler with -v -E on an no-op compilation unit and parses the built-in
+    include search directories and the GCC version from the output
+
+    :param compiler: name / path of the compiler to run
+    :type compiler: str
+
+    :return: (list_of_include_paths, version)
+    :rtype: tuple
+    """
+    try:
+        with subprocess.Popen([compiler, "-v", "-E", "-"],
+                              stdin=subprocess.PIPE, stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE) as proc:
+            inputdata = b"typedef int dont_be_pedantic;"
+            _, stderrdata = proc.communicate(input=inputdata)
+    except FileNotFoundError:
+        msg = "Compiler {} not found, not adding system include paths\n".format(compiler)
+        sys.stderr.write(msg)
+        return []
+
+    stderrdata = stderrdata.decode("utf-8")
+    version = REGEX_VERSION.search(stderrdata).group(1)
+    includes = [os.path.abspath(p) for p in REGEX_INCLUDES.search(stderrdata).group(1).split()]
+
+    return (includes, version)
+
+
+def detect_libstdcxx_includes(compiler, includes, version):
+    """
+    Tries to detect the g++ libstdc++ built-in include search directories using black magic and
+    adds them to the list given in includes
+
+    :param compiler: Name or path of the compiler
+    :type compiler: str
+    :param includes: List of include directories
+    :type includes: list of str
+    :param version: Version of g++
+    :type version: str
+    """
+    for path in includes:
+        cxx_lib = os.path.join(path, "c++", version)
+        if os.path.exists(cxx_lib):
+            includes.append(cxx_lib)
+            triple = os.path.basename(compiler)[0:-4]
```

Comment:
Are there plans for one GCC to rule them all? That would make packaging cross compilers much easier :-)

Regarding the question: On the one hand, this will no longer work. (But the path with the triple seems to only be present on arm-none-eabi anyway, so I could just test for this hard coded path instead.) On the other hand, I kind of expect paths to become explicit arguments like they already are for `clang` when compiling for the non-default target triple. If that is true, this hole dark magic code could be dropped for good :-)